### PR TITLE
Check for SIGHASH_SINGLE bug in writer fn

### PR DESCRIPTION
**Pull request by tcharding**
Recently we moved the logic for checking for the SIGHASH_SINGLE bug to
the `signature_hash()` function. Although this left users of the
`encode_signing_data_to()` function without correct handling of the bug
there is not much else we can do but alert users to this behaviour.

Add documentation to highlight the behaviour of `encdoe_signing_data_to`
in regards to the sighash single bug. Requires updating docs for
`signature_hash` also.

Please note, uses non-conventional markdown header `# Warning`.

Closes: #817

### CODE REVIEW

**Review by dr-orlovsky — COMMENTED** [review]
Are we sure that the right way here is to error and not panic?

**Comment by apoelstra** [comment]
On further reflection I think we should just document that this function does the wrong thing with the SIGHASH_SINGLE bug, and that if the user intends to care about this, they should either call `signature_hash` or write their own logic.

Changing an infallible function to a fallible one, without changing the type signature, seems like a bad change to make to the library.

**Comment by elichai** [comment]
I might've asked this already on IRC and I don't remember the answer.
What's the use case of calling `encode_signing_data_to` and not `signature_hash`?
if there's no real use case then the best solution IMHO is to make this function private

**Comment by apoelstra** [comment]
I think the original author https://github.com/rust-bitcoin/rust-bitcoin/pull/485 just made it public and nobody noticed or complained. I don't remember explicitly thinking about this ever.

**Comment by apoelstra** [comment]
Ah, great, thanks for finding this.

So -- if we added an error case, and the author has `.expect()`ed the output of this function, in theory this could result in his HSM being crashable from the outside, by a user providing a transaction that tries to use the `SIGHASH_SINGLE` bug. Contrast that with the current behavior where the HSM will harmlessly sign a bad sighash.

So I think we should just document "does the wrong thing given the sighash single bug". Maybe we should even make this explicit, by feeding `#"[not a transaction] SIGHASH_SINGLE bug"` into the writer or something.

**Comment by tcharding** [comment]
Total re-write of PR including update of PR description. I built the docs and they look ok to me but please not the `Warning` heading is non-conventional.

File: src/blockdata/transaction.rs
Comment (sanket1729) [inline src/blockdata/transaction.rs:356]: Either this should be returning an Err(encode::Error::SigHashSingle) or the error enum variant should be removed.

Comment (tcharding) [reply src/blockdata/transaction.rs:356]: Thanks man, sloppy work by me leaving that in - no excuse.

**Review by sanket1729 — APPROVED** [review]
ACK 83dda74ecb7e00c1a4e5726a75e4769079e588e8. This is much cleaner

File: src/blockdata/transaction.rs
Comment (sanket1729) [reply src/blockdata/transaction.rs:356]: No worries, happens to me a lot :)

**Comment by dr-orlovsky** [comment]
ACK 83dda74ecb7e00c1a4e5726a75e4769079e588e8

**Review by dr-orlovsky — APPROVED** [review]
ACK 83dda74ecb7e00c1a4e5726a75e4769079e588e8

Not merging: waiting for @apoelstra approval as well, since he was the main person addressing this issue in the previous PR version and elsewhere

**Review by apoelstra — APPROVED** [review]
ACK 83dda74ecb7e00c1a4e5726a75e4769079e588e8

Thank you for waiting for me!

**Pull request merged**
